libflux/src/flux/lib.rs

```diff
@@ -458,31 +505,77 @@ pub unsafe extern "C" fn flux_analyze_with(
 /// analyze consumes the given AST package and returns a semantic package
 /// that has been type-inferred.  This function is aware of the standard library
 /// and prelude.
-pub fn analyze(ast_pkg: ast::Package) -> Result<core::semantic::nodes::Package, core::Error> {
+pub fn analyze(ast_pkg: ast::Package) -> Result<Package, Error> {
+    let (sem_pkg, _, sub) = infer_with_env(ast_pkg, fresher(), None)?;
+    Ok(inject_pkg_types(sem_pkg, &sub))
+}
+
+/// infer_with_env consumes the given AST package, inject the type bindings from the given
+/// type environment, and returns a semantic package that has not been type-injected and an
+/// inferred type environment and substitution.
+/// This function is aware of the standard library and prelude.
+pub fn infer_with_env(
+    ast_pkg: ast::Package,
+    mut f: Fresher,
+    env: Option<Environment>,
+) -> Result<(Package, Environment, Substitution), Error> {
     // First check to see if there are any errors in the AST.
     let errs = ast::check::check(ast::walk::Node::Package(&ast_pkg));
     if !errs.is_empty() {
         return Err(core::Error::from(format!("{}", &errs[0])));
     }
 
     let pkgpath = ast_pkg.path.clone();
-    let mut f = fresher();
     let mut sem_pkg = core::semantic::convert::convert_with(ast_pkg, &mut f)?;
 
     check::check(&sem_pkg)?;
 
-    let prelude = match prelude() {
+    let mut prelude = match prelude() {
         Some(prelude) => Environment::new(prelude),
         None => return Err(core::Error::from("missing prelude")),
     };
+    if let Some(e) = env {
+        prelude.copy_bindings_from(&e);
+    }
     let imports = match imports() {
         Some(imports) => imports,
         None => return Err(core::Error::from("missing stdlib imports")),
     };
     let builtin_importer = builtins().importer_for(&pkgpath, &mut f);
-    let (_, sub) = infer_pkg_types(&mut sem_pkg, prelude, &mut f, &imports, &builtin_importer)?;
-    sem_pkg = inject_pkg_types(sem_pkg, &sub);
-    Ok(sem_pkg)
+
+    let (env, sub) = infer_pkg_types(&mut sem_pkg, prelude, &mut f, &imports, &builtin_importer)?;
+    Ok((sem_pkg, env, sub))
+}
+
+/// Given a Flux source and a variable name, find out the type of that variable in the Flux source code.
+/// A type variable will be automatically generated and injected into the type environment that
+/// will be used in semantic analysis. The Flux source code itself should not contain any definition
+/// for that variable.
+/// This version of find_var_type is aware of the prelude and builtins.
+pub fn find_var_type(ast_pkg: ast::Package, var_name: String) -> Result<MonoType, Error> {
```

Comment:
I am confused why there are two `find_var_type` functions? When is one or the other used?

Reply:
The one in the `flux` crate is aware of the prelude and the one in `core` is not.
The one in `core` is just for testing purpose